packages/react-devtools-extensions/src/parseHookNames.js

```diff
@@ -0,0 +1,501 @@
+        if (runtimeSourceCode.length > MAX_SOURCE_LENGTH) {
+          throw Error('Source code too large to parse');
+        }
+
+        if (__DEBUG__) {
+          console.groupCollapsed(
+            'loadSourceFiles() fileName "' + fileName + '"',
+          );
+          console.log(runtimeSourceCode);
+          console.groupEnd();
+        }
+
+        hookSourceData.runtimeSourceCode = runtimeSourceCode;
+      }),
+    );
+  });
+  return Promise.all(promises);
+}
+
+async function parseSourceAST(
+  fileNameToHookSourceData: Map<string, HookSourceData>,
+): Promise<*> {
+  // SourceMapConsumer.initialize() does nothing when running in Node (aka Jest)
+  // because the wasm file is automatically read from the file system
+  // so we can avoid triggering a warning message about this.
+  if (!__TEST__) {
+    if (__DEBUG__) {
+      console.log('parseSourceAST() Initializing source-map library ...');
+    }
+
+    // $FlowFixMe
+    const wasmMappingsURL = chrome.extension.getURL('mappings.wasm');
+
+    SourceMapConsumer.initialize({'lib/mappings.wasm': wasmMappingsURL});
+  }
+
+  const promises = [];
+  fileNameToHookSourceData.forEach(hookSourceData => {
+    if (hookSourceData.originalSourceAST !== null) {
+      // Use cached metadata.
+      return;
+    }
+
+    const {runtimeSourceCode, sourceMapContents} = hookSourceData;
+    if (sourceMapContents !== null) {
+      // Parse and extract the AST from the source map.
+      promises.push(
+        SourceMapConsumer.with(
+          sourceMapContents,
+          null,
+          (sourceConsumer: SourceConsumer) => {
+            hookSourceData.sourceConsumer = sourceConsumer;
+
+            // Now that the source map has been loaded,
+            // extract the original source for later.
+            const source = sourceMapContents.sources[0];
+            const originalSourceCode = sourceConsumer.sourceContentFor(
+              source,
+              true,
+            );
+
+            if (__DEBUG__) {
+              console.groupCollapsed(
+                'parseSourceAST() Extracted source code from source map',
+              );
+              console.log(originalSourceCode);
+              console.groupEnd();
+            }
+
+            hookSourceData.originalSourceCode = originalSourceCode;
+
+            // TODO Parsing should ideally be done off of the main thread.
+            hookSourceData.originalSourceAST = parse(originalSourceCode, {
+              sourceType: 'unambiguous',
+              plugins: ['jsx', 'typescript'],
+            });
+          },
+        ),
+      );
+    } else {
+      // There's no source map to parse here so we can just parse the original source itself.
+      hookSourceData.originalSourceCode = runtimeSourceCode;
+
+      // TODO Parsing should ideally be done off of the main thread.
+      hookSourceData.originalSourceAST = parse(runtimeSourceCode, {
+        sourceType: 'unambiguous',
+        plugins: ['jsx', 'typescript'],
```

Comment:
A non-perf reason to invest in #21782 ( = derive metadata from the AST at build time) is that there's no single set of plugins we can pass here that will work for all use cases.

1. If the source code is written in Flow, it likely will not parse as TypeScript.
2. People using experimental syntax in their code will run into parse errors here. There used to be a catch-all `'*'` option you could pass to Babel and enable all plugins, but that was removed when plugins with conflicting semantics were introduced (Flow / TypeScript, divergent proposals for decorators/pipeline operators etc)

Reply:
Another good point. I also wonder about perf and file size impact for enabling all plugins, even if that were still an option.